### PR TITLE
Support skeleton formats

### PR DESCRIPTION
**Pull request by c960657**
In addition to the `full`, `long`, `medium` and `short` formats, CLDR defines a number of [additional formats](http://unicode.org/cldr/trac/browser/trunk/common/main/en.xml#L2211).

This patch adds a method  `getSkeletonFormat()` to look up the associated pattern. It also allows skeletons to be passed to functions accepting accepting a width (`full` etc.) by prefixing the skeleton with `~`, e.g. `Calendar::formatDate($dt1, '~GyMMMEd')`.

The user must supply one of the skeleton formats specified explicitly in the CLDR data files. The matching algorithm described in [UTS #35, part 4, section 2.6.2.1](http://www.unicode.org/reports/tr35/tr35-dates.html#Matching_Skeletons) has not (yet) been implemented, but it may be in the future without changing the API.

I have moved the logic in `getDatetimeFormatReal()` for determining `$wholeWidth` when two chuncks are specified to a separate function, `getDatetimeWidth()`. I have changed the existing logic to be more in line with the algorithm described in [UTS #35, part 4, section 2.6.2.2](http://www.unicode.org/reports/tr35/tr35-dates.html#Missing_Skeleton_Fields), where the date pattern determines the whole width pattern to use.

### CODE REVIEW

File: code/Calendar.php [hunk 1]
```diff
@@ -1015,6 +1016,10 @@ public static function getSortedWeekdays($namesWidth = false, $locale = '')
      */
     public static function getDateFormat($width, $locale = '')
     {
+        if (substr($width, 0, 1) === '~') {
```
Comment (mlocati) [inline code/Calendar.php]: What about `($width !== '' && $width[0] === '~')` ? It's ways faster...

**Comment by mlocati** [comment]
I like this idea!

File: code/Calendar.php (hunk 1 above)
Comment (c960657) [inline code/Calendar.php]: Fixed. FWIW, I could reproduce the performance issue in PHP 5 and 7.0, but in PHP 7.1 there does not seem to be any performance difference.

**Comment by c960657** [comment]
I have now added support for non-exact matching of skeletongs, millisecond handling and support for input skeleton symbols (j, J, C).

The non-exact matching still has some limitations (see comment for `getBestMatchingSkeleton()`), but I think it is still useful in its current state.

The spec is pretty vague/ambiguous on some of these topics, so I have used ICU4j as a (somewhat) authoritative source on how to interpret certain things. I used the tool on http://demo.icu-project.org/icu4jweb/flexTest.jsp to verify my results.


`getBestMatchingSkeleton()` will also be useful for matching `intervalFormatItem`s. I expect to create a separate PR for this feature.

**Comment by c960657** [comment]
FYI, my work on interval formatting is progressing here: [c960657:format-interval](https://github.com/c960657/punic/compare/skeleton-format...c960657:format-interval). I suggest this PR is not merged until both features are finished.

**Comment by mlocati** [comment]
> I suggest this PR is not merged until both features are finished.

No problems at all: just tell me when it's ready

**Comment by c960657** [comment]
I am pretty satisfied with this now and with the interval formatting in the [format-interval branch](https://github.com/c960657/punic/compare/skeleton-format...c960657:format-interval) (I will create a separate PR for that once this one has landed).

**Comment by mlocati** [comment]
@c960657 Thanks for the great work!
BTW we have a breaking change in `\Punic\Calendar::getDatetimeFormat()`...

For instance, as highlighted by [this change](https://github.com/punic/punic/pull/92/files#diff-aa2c8fabe7931a5af5c31858071ebe7cL819), let's assume we have `Calendar::getDatetimeFormat('short|full')`.
Before we took the maximum width (so, `full` in the example), now we take the date width (so, `short` in the example)

**Comment by c960657** [comment]
The change in `Calendar::getDatetimeFormat('short|full')` was actually intentional in order to become more in line with the logic described in the CLDR spec, i.e. let `$wholeWidth` be determined by the date field only.

The behaviour of the function is changed, but it does not violate the documented behaviour (the Docblock comment does not specify which whole-width format is to be used if none is specified explicitly).

Anyway, I understand your concern. Do you think I should restore the old behaviour for full/long/medium/short?

**Comment by mlocati** [comment]
> Do you think I should restore the old behaviour for full/long/medium/short?

I'm uncertain about this: by one side I'd prefer your approach (using the width of the date), but it's a breaking change.
But since I think it's just a minor BC, let's keep your approach :+1:

**Comment by mlocati** [comment]
@c960657 This is a really great work! Thanks!!

**Pull request merged**